text/0000-impl-trait-type-aliases.md

```diff
@@ -0,0 +1,414 @@
+}
+```
+
+In this example, the concrete type referred to by `Baz` is guaranteed to be the same wherever `Baz` occurs.
+
+Note that using `Baz` as an argument type is *not* the same as argument-position `impl Trait`, as `Baz` refers to a unique type, whereas the concrete type for argument-position `impl Trait` is determined by the caller.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+fn foo(x: Baz) {
+    // ...
+}
+
+// is *not* the same as:
+
+fn foo(x: impl Trait) {
+    // ...
+}
+```
+
+Just like with any other type alias, we can use `impl Trait` to specify associated types for traits, as in the following example.
+
+```rust
+trait Trait {
+    type Assoc;
+}
+
+struct Foo {}
+
+impl Trait for Foo {
+    type Assoc = impl Debug;
+}
+```
+
+Here, anything that makes use of `Foo` knows that `Foo::Assoc` implements `Debug`, but has no knowledge of its concrete type.
+
+[Eventually, we would also describe the use of `impl Trait` in `let`, `const` and `static` bindings, but as they are as-yet unimplemented and function the same as return-type `impl Trait`, they haven't been included here.]
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Since RFC 2071 was accepted, the initial implementation of `existential type` [has already been completed](https://github.com/rust-lang/rust/pull/52024). This RFC would replace the syntax of `existential type`, from:
+
+```rust
+existential type Foo: Bar;
+```
+
+to:
+
+```rust
+type Foo = impl Bar;
+```
+
+In addition, having multiple occurrences of `impl Trait` in a type alias or associated type is now permitted, where each occurrence is desugared into a separate inferred type. For example, the following alias:
+
+```rust
+type Foo = Arc<impl Iterator<Item = impl Debug>>;
+```
+
+would be desugared to the equivalent of:
+
+```rust
+existential type _0: Debug;
+existential type _1: Iterator<Item = _0>;
+type Foo = Arc<_1>;
+```
+
+Furthermore, when documenting `impl Trait`, explanations of the feature would avoid type theoretic terminology (specifically "existential types") and prefer type inference language (if any technical description is needed at all).
+
+`impl Trait` type aliases may contain generic parameters just like any other type alias. The type alias must contain the same type parameters as its concrete type, except those implicitly captured in the scope (see [RFC 2071](https://github.com/rust-lang/rfcs/blob/master/text/2071-impl-trait-existential-types.md) for details).
+
+```rust
+// `impl Trait` type aliases may contain type parameters...
+#[derive(Debug)]
+struct DebugWrapper<T: Debug>(T);
+
+type Foo<T> = impl Debug;
+
+fn get_foo<T: Debug>(x: T) -> Foo<T> { DebugWrapper(x) }
+
+// ...and lifetime parameters (and so on).
+#[derive(Debug)]
+struct UnitRefWrapper<'a>(&'a ());
+
+type Bar<'a> = impl Debug;
+
+fn get_bar<'a>(y: &'a ()) -> Bar<'a> { UnitRefWrapper(y) }
+```
+
+## Restricting compound `impl Trait` trait aliases
```

Comment:
On the language team meeting we noticed that you forgot to remove this section since it now conflicts with the description above.